### PR TITLE
Add parsing code for the Font Features, as well as new command

### PR DESCRIPTION
**Pull request by bramp**

### CODE REVIEW

**Comment by bramp** [comment]
For example:
```
$ go run main.go features ~/path/to/font.woff 

Glyph Substitution Table (GSUB):
	Script "DFLT" (Default):
		Default Language:
			Feature "aalt" (Access All Alternates)
			Feature "c2sc" (Small Capitals From Capitals)
			Feature "case" (Case-Sensitive Forms)
			Feature "crcy" (Currency (Deprecated))
		Language "AZE " (Azerbaijani):
			Feature "smcp" (Small Capitals)
	Script "cyrl" (Cyrillic):
		Default Language:
			Feature "aalt" (Access All Alternates)
			Feature "c2sc" (Small Capitals From Capitals)
			Feature "case" (Case-Sensitive Forms)
			Feature "crcy" (Currency (Deprecated))

Glyph Positioning Table (GPOS):
	Script "DFLT" (Default):
		Default Language:
			Feature "kern" (Kerning)
	Script "cyrl" (Cyrillic):
		Default Language:
			Feature "kern" (Kerning)
	Script "latn" (Latin):
		Default Language:
			Feature "kern" (Kerning)
		Language "AZE " (Azerbaijani):
			Feature "kern" (Kerning)
```

File: commands/features.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package commands
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/ConradIrwin/font/sfnt"
+)
+
+func layoutTable(font *sfnt.Font, tag sfnt.Tag, name string) {
```
Comment (dmitshur) [inline commands/features.go]: I’d suggest moving `layoutTable` to the bottom, under `Features`, rather than on top.

In Go, it’s common to order code from higher level functions first to lower level helpers second. That way, when reading code, you first encounter higher level, easier to understand code that helps provide context for the helpers. Then, once you're familiar with how a helper is being used and why it’s needed, reading its code is easier.

**Comment by dmitshur** [comment]
(The rest of this PR is awesome, nice work!)

File: commands/features.go (hunk 1 above)
Comment (bramp) [reply commands/features.go]: Thanks for the feedback. Good point. Fixed.

File: commands/features.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package commands
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/ConradIrwin/font/sfnt"
+)
+
+// Features prints the gpos/gsub tables (contins font features).
```
Comment (dmitshur) [inline commands/features.go]: s/contins/contains/

File: main.go [hunk 3]
```diff
@@ -25,14 +25,17 @@ func main() {
...
 	case "metrics":
 		commands.Metrics()
+	case "features":
+		commands.Features()
 	default:
 		fmt.Println(`
-Usage: font [scrub|info|stats] font.[otf,ttf,woff]
+Usage: font [features|info|metrics|scrub|stats] font.[otf,ttf,woff]
 
+features: prints the gpos/gsub tables (contins font features)
```
Comment (dmitshur) [inline main.go]: s/contins/contains/

**Review by dmitshur — COMMENTED** [review]
Spotted a couple typos.

File: commands/features.go (hunk 2 above)
Comment (bramp) [reply commands/features.go]: fixed

File: main.go (hunk 3 above)
Comment (bramp) [reply main.go]: fixed

File: sfnt/table_feature.go [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+
+// Script returns the name for this feature.
+func (f *Feature) String() string {
+	tag := f.Tag.String()
+	if name, found := featureTags[tag]; found {
+		return name
+	}
+
+	if len(tag) == 4 && tag[0] == 'c' && tag[1] == 'v' && tag[2] >= '0' && tag[2] <= '9' && tag[3] >= '0' && tag[3] <= '9' {
+		return "Character Variants"
```
Comment (ConradIrwin) [inline sfnt/table_feature.go]: @bramp might be worth preserving the number here (e.g. Character Variables 01)

File: commands/features.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	"fmt"
+	"os"
+
+	"github.com/ConradIrwin/font/sfnt"
+)
+
+// Features prints the gpos/gsub tables (contains font features).
+func Features() {
+	if len(os.Args) < 2 {
+		panic(fmt.Errorf("Specify a font file"))
```
Comment (ConradIrwin) [inline commands/features.go:13]: Can we start to improve the code for handling these errors, we should make it clear how to fix the problem, and make the error message more legible. (I appreciate the other commands aren't yet very good).

```
fmt.Printf("Usage: font features <font-file>")
os.Exit(1)
```

File: commands/features.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+	layoutTable(font, sfnt.TagGsub, "Glyph Substitution Table (GSUB)")
+	layoutTable(font, sfnt.TagGpos, "Glyph Positioning Table (GPOS)")
+}
+
+func layoutTable(font *sfnt.Font, tag sfnt.Tag, name string) {
+	if font.HasTable(tag) {
+		fmt.Printf("%s:\n", name)
+
+		t := font.Table(tag).(*sfnt.TableLayout)
```
Comment (ConradIrwin) [inline commands/features.go:35]: @bramp This will panic if parsing fails. I'm not sure what to do about that except maybe change the library so that it parses tables on-demand instead of up-front so we can return a sensible error. Any better ideas?

File: commands/features.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+		for _, script := range t.Scripts {
+			fmt.Printf("\tScript %q (%s):\n", script.Tag, script.String())
```
Comment (ConradIrwin) [inline commands/features.go]: When I run this on a font I see a weird:

`Script "    " ():`

Do you know why? Can we make it clearer what it means?

File: sfnt/table_feature.go (hunk 4 above)
Comment (ConradIrwin) [reply sfnt/table_feature.go]: (Also, should we do the same parsing for `ss01`...? )

File: sfnt/table_feature.go [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+}
+type scriptRecord tagOffsetRecord
+type featureRecord tagOffsetRecord
+type lookupRecord tagOffsetRecord
+type langSysRecord tagOffsetRecord
+
+type scriptTable struct {
+	DefaultLangSys uint16 // Offset to default LangSys table, from beginning of Script table — may be NULL
+	LangSysCount   uint16 // Number of LangSysRecords for this script — excluding the default LangSys
+	// langSysRecords[langSysCount] langSysRecord // Array of LangSysRecords, listed alphabetically by LangSys tag
```
Comment (ConradIrwin) [inline sfnt/table_feature.go:123]: @bramp Why is this here but commented out? Should we uncomment it or delete it?

File: sfnt/table_feature.go [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+}
+
+// parseScript parses a single Script table. b expected to be the beginning of ScriptList.
+// See https://www.microsoft.com/typography/otspec/chapter2.htm#sTbl_lsRec
+func (t *TableLayout) parseScript(b []byte, record scriptRecord) (*Script, error) {
+	if int(record.Offset) >= len(b) {
+		return nil, io.ErrUnexpectedEOF
+	}
+
+	b = b[record.Offset:]
```
Comment (ConradIrwin) [inline sfnt/table_feature.go:187]: @bramp Is there a byte length on the record? I'm worried about the parsing code reading beyond the length of the record.

**Comment by ConradIrwin** [comment]
Otherwise this looks great.

If you have the time, a test would be great, but happy to merge without.

**Comment by ConradIrwin** [comment]
(Thanks for the other view @shurcooL )

File: commands/features.go (hunk 7 above)
Comment (bramp) [reply commands/features.go]: Would you email me that font?

File: sfnt/table_feature.go (hunk 4 above)
Comment (bramp) [reply sfnt/table_feature.go]: Sure, will fix.

**Comment by ConradIrwin** [comment]
P.S I added you both as collaborators to this repository.

Please notify me for PRs with large changes or API changes — but happy if you want to merge your own PRs for tweaks to code-style / error messages.

File: commands/features.go (hunk 7 above)
Comment (ConradIrwin) [reply commands/features.go]: it's here: https://gist.github.com/ConradIrwin/a85f8e432fedf2cd4ea400286a796b5e 

If you have the gist tool installed (brew install gist) you can get the file with: `gist -r a85f8e432fedf2cd4ea400286a796b5e | base64 --decode > adelle.ttf`

File: commands/features.go (hunk 6 above)
Comment (bramp) [reply commands/features.go:35]: Yes, panic'ng sucks. After this I'd be happy to change the whole library to pass back the errors.

Comment (ConradIrwin) [reply commands/features.go:35]: That'd be incredible 😻

File: sfnt/table_feature.go (hunk 4 above)
Comment (bramp) [reply sfnt/table_feature.go]: Added numeric suffix, plus added support for ss01... and added a test.

File: commands/features.go (hunk 5 above)
Comment (bramp) [reply commands/features.go:13]: I'll do that in a later PR.

File: sfnt/table_feature.go (hunk 8 above)
Comment (bramp) [reply sfnt/table_feature.go:123]: Ah yes, perhaps unclear. The ondisk format has a variable length array as the last element. I explicitly deal with that in the code, but left the field commented to indicate it was there.

I can drop the comment, or make it clear why it's there.

File: sfnt/table_feature.go (hunk 9 above)
Comment (bramp) [reply sfnt/table_feature.go:187]: nope, there is sadly no length for any of these fields. I suspect they are all contained within the single table.

File: commands/features.go (hunk 7 above)
Comment (bramp) [reply commands/features.go]: So that font has a script who's tag is 0x20202020 (four space characters).  The spec says nothing about what this means, so I suspect it is just some weird artefact.

I've changed the tool to not display the () if the name is unknown.

**Comment by bramp** [comment]
I've pushed my changes, if you want to make one more pass before merging?

File: sfnt/table_feature.go (hunk 9 above)
Comment (bramp) [reply sfnt/table_feature.go:187]: ah yes, all the records are within the same table, and that b array only contains the current table:

```go
buffer := io.NewSectionReader(file, int64(entry.Offset), int64(entry.Length))
table, err := parseTable(entry.Tag, buffer)
...
    b, err := ioutil.ReadAll(buffer)
...
    b = b[record.Offset:]

**Comment by ConradIrwin** [comment]
Thanks!
